frontend/src/modules/fluenteditor/components/FluentEditor.js

```diff
@@ -0,0 +1,281 @@
+    );
+    const locale = useSelector((state) => state.locale);
+
+    // Force using the source editor.
+    const [forceSource, setForceSource] = React.useState(false);
+
+    // When the entity changes, reset the `forceSource` setting. Never show the source
+    // editor by default.
+    React.useEffect(() => {
+        setForceSource(false);
+    }, [entity]);
+
+    // When a user wants to force (or unforce) the source editor, we need to convert
+    // the existing translation to a format appropriate for the next editor type.
+    function changeForceSource() {
+        if (syntaxType === 'complex') {
+            return;
+        }
+        const fromSyntax = forceSource ? 'complex' : syntaxType;
+        const toSyntax = forceSource ? syntaxType : 'complex';
+        const [translationContent, initialContent] = fluent.convertSyntax(
+            fromSyntax,
+            toSyntax,
+            translation,
+            entity.original,
+            activeTranslationString,
+            locale,
+        );
+        dispatch(editor.actions.setInitialTranslation(initialContent));
+        dispatch(editor.actions.update(translationContent));
+        setForceSource(!forceSource);
+    }
+
+    return [forceSource, changeForceSource];
+}
+
+/**
+ * Editor dedicated to modifying Fluent strings.
+ *
+ * Renders the most appropriate type of editor for the current translation.
+ */
+export default function FluentEditor() {
+    const dispatch = useDispatch();
+
+    const isReadOnlyEditor = useSelector((state) =>
+        entities.selectors.isReadOnlyEditor(state),
+    );
+    const user = useSelector((state) => state.user);
+
+    const syntaxType = useAnalyzeSyntax();
+    const [forceSource, changeForceSource] = useForceSource(syntaxType);
+    useLoadTranslation(syntaxType, forceSource);
+
+    // Do not render if the syntax has not yet been computed.
+    if (syntaxType === '') {
+        return null;
+    }
+
+    // Choose which editor implementation to render.
+    let EditorImplementation = RichEditor;
+    if (forceSource || syntaxType === 'complex') {
+        EditorImplementation = SourceEditor;
+    } else if (syntaxType === 'simple') {
+        EditorImplementation = SimpleEditor;
+    }
+
+    // When the syntax is complex, the editor is blocked in source mode, and it
+    // becomes impossible to switch to a different editor type. Thus we show a
+    // notification to the user if they try to use the "FTL" switch button.
+    function showUnsupportedMessage() {
+        dispatch(
+            notification.actions.add(
+                notification.messages.FTL_NOT_SUPPORTED_RICH_EDITOR,
+            ),
+        );
+    }
+
+    // Show a button to allow switching to the source editor.
+    let ftlSwitch = null;
+    // But only if the user is logged in and the string is not read-only.
+    if (user.isAuthenticated && !isReadOnlyEditor) {
+        if (syntaxType === 'complex') {
+            // TODO: To Localize
```

Comment:
Were you planning to do the localization here and on #L267 in this patch?

Reply:
No, that's outside the scope of this, which is why I put those comments instead. I think there are still a few parts that are not localized throughout the app, and we'll need to make a thorough audit of the app before enabling translations. But that should be left for a different bug! :-)